### PR TITLE
DISPATCH-2338 - dockerfiles: Introduce Debian-based slim image w/ CI tests

### PR DESCRIPTION
**Pull request by Jmennius**
Use Debian stable release slim image with multistage build to
keep image size reasonable (143MB compared to 1.39GB Ubuntu based image).

Other changes compared to Ubuntu based image:
* ~~libuv is enabled~~
* a bit more careful work with dependencies
* allow building web console
* faster builds with Ninja
* build Proton from release, not master; configurable
* build Dispatch Router from checked out source
* minor changes in build process, Dockerfile syntax, etc

TODO:
- [x] build image on CI

---
It's not the prettiest in some ways.. but what can I do :)

~~I did not run any real workloads with it yet, but the router daemon starts and qdmanage can query from it.~~ tested

### CODE REVIEW

File: dockerfiles/Dockerfile-debian
Comment (jiridanek) [inline dockerfiles/Dockerfile-debian:42]: Why libuv and not the default epoll?

File: dockerfiles/Dockerfile-debian
Comment (jiridanek) [inline dockerfiles/Dockerfile-debian:74]: Why is this written as `COPY` first, `RUN apt-get` afterwards. I thought that the opposite order is more reasonable.

File: dockerfiles/Dockerfile-debian
Comment (jiridanek) [inline dockerfiles/Dockerfile-debian:47]: Is Ninja really such a big win? I do use it interactively, but mostly because it is fast on rebuilds. Using make with `-j` option for first build is IMO quite comparable in speed with ninja.

I'm fine with ninja in that it avoids having to mess with the `-j` make option.

File: dockerfiles/Dockerfile-debian
Comment (jiridanek) [inline dockerfiles/Dockerfile-debian:77]: Yeah, the `site-packages` bug is really annoying...

File: dockerfiles/Dockerfile-debian
Comment (jiridanek) [inline dockerfiles/Dockerfile-debian:27]: :+1:

**Review by jiridanek — COMMENTED** [review]
I like it, overall. I have a few comments inline.

File: dockerfiles/Dockerfile-debian
Comment (Jmennius) [reply dockerfiles/Dockerfile-debian:42]: Hard to say, I am not opposed to either one - I just though this was the intention of other images since they installed it (but did not enable).

File: dockerfiles/Dockerfile-debian
Comment (Jmennius) [reply dockerfiles/Dockerfile-debian:47]: Yes, that was exactly the reason to use it - to avoid hard coding or messing with `-j`

File: dockerfiles/Dockerfile-debian
Comment (Jmennius) [reply dockerfiles/Dockerfile-debian:74]: It probably is, I will change it. (I will also include `ca-certificates` in the final image)

File: dockerfiles/Dockerfile-debian
Comment (Jmennius) [reply dockerfiles/Dockerfile-debian:77]: yeah, but minor enough for me to ignore it for now :)

**Comment by Jmennius** [comment]
In case you are willing to accept this - should I add build to CI?

**Comment by jiridanek** [comment]
> In case you are willing to accept this - should I add build to CI?

Yes, that would be quite helpful. In fact, we should probably build all the dockerfiles. But building just this one will be a good start.

**Comment by Jmennius** [comment]
> > In case you are willing to accept this - should I add build to CI?
> 
> Yes, that would be quite helpful. In fact, we should probably build all the dockerfiles. But building just this one will be a good start.

Ok, I'll do that a little bit later.
I have pushed a commit with requested changes (lets squash that later).

File: dockerfiles/Dockerfile-debian
Comment (gemmellr) [inline dockerfiles/Dockerfile-debian:38]: I'd suggest using the GitHub repo, especially if going to build it in CI.

File: dockerfiles/Dockerfile-debian
Comment (Jmennius) [reply dockerfiles/Dockerfile-debian:78]: I meant to ignore it but it's Monday and I was curious...
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=998739

**Comment by jiridanek** [comment]
https://github.com/apache/qpid-dispatch/runs/5980336959?check_suite_focus=true#step:2:96

```
/usr/bin/docker buildx build --file dockerfiles/Dockerfile-debian --iidfile /tmp/docker-build-push-8nmE0m/iidfile --metadata-file /tmp/docker-build-push-8nmE0m/metadata-file .
error: could not find dockerfiles: stat dockerfiles: no such file or directory
Error: buildx failed with: error: could not find dockerfiles: stat dockerfiles: no such file or directory
```

The image build job looks broken

**Comment by Jmennius** [comment]
> https://github.com/apache/qpid-dispatch/runs/5980336959?check_suite_focus=true#step:2:96
> 
> ```
> /usr/bin/docker buildx build --file dockerfiles/Dockerfile-debian --iidfile /tmp/docker-build-push-8nmE0m/iidfile --metadata-file /tmp/docker-build-push-8nmE0m/metadata-file .
> error: could not find dockerfiles: stat dockerfiles: no such file or directory
> Error: buildx failed with: error: could not find dockerfiles: stat dockerfiles: no such file or directory
> ```
> 
> The image build job looks broken

Let's try again... (needs approval)

File: dockerfiles/Dockerfile-debian
Comment (jiridanek) [reply dockerfiles/Dockerfile-debian:82]: Ticket at https://issues.apache.org/jira/browse/DISPATCH-2337, to keep track of the defect.

**Comment by Jmennius** [comment]
> > Let's try again... (needs approval)
> 
> It should be running without approvals on your GH fork, so you should already know if success is expected. I don't know why you don't have the GHA running at https://github.com/Jmennius/qpid-dispatch/actions.

That steps builds now :)

**Comment by jiridanek** [comment]
@ganeshmurthy @kgiusti I think this looks ok, and it has the benefit of coming with a CI job. Furthermore the image is smaller than what can be achieved with CentOS bases (but ubi8 should now work to produce small image, if we either decide to live without fetching in gdb ;P or pull it from CentOS).

In general I'd like to delete some of the dockerfiles, maybe all except one, and leave the Debian one to be the only one Containerfile in `main` branch, WDYT?

**Comment by Jmennius** [comment]
Few additional questions
1. should be web console be enabled by default?
2. we're fine with `RelWithDebInfo` (default) or should we switch to `Release`?
3. is it worth adding `$(if [ "${enable_console}" = "ON" ]; then echo "npm"; fi)` to apt-get invocation on build stage to avoid installing `npm` when it is not required during build?

**Comment by jiridanek** [comment]
> 1. should be web console be enabled by default?

imo no

> 2. we're fine with `RelWithDebInfo` (default) or should we switch to `Release`?

RelWithDebInfo is used everywhere, nobody ever uses Release with dispatch, as far as I know

> 3. is it worth adding `$(if [ "${enable_console}" = "ON" ]; then echo "npm"; fi)` to apt-get invocation on build stage to avoid installing `npm` when it is not required during build? (does not influence runtime image)

maybe, yes. If you put it on its own line, then it makes sense to me; of course, it complicates build... up to you, I think

**Comment by Jmennius** [comment]
> > 1. should be web console be enabled by default?
> 
> imo no
> 
> > 2. we're fine with `RelWithDebInfo` (default) or should we switch to `Release`?
> 
> RelWithDebInfo is used everywhere, nobody ever uses Release with dispatch, as far as I know
> 
> > 3. is it worth adding `$(if [ "${enable_console}" = "ON" ]; then echo "npm"; fi)` to apt-get invocation on build stage to avoid installing `npm` when it is not required during build? (does not influence runtime image)
> 
> maybe, yes. If you put it on its own line, then it makes sense to me; of course, it complicates build... up to you, I think

Thanks. If console stays disabled by default - I'll add that condition then.

**Comment by jiridanek** [comment]
I've reported the failing CentOS 7 GHA job as issue https://issues.apache.org/jira/browse/PROTON-2534

File: .github/workflows/build.yaml
Comment (jiridanek) [inline .github/workflows/build.yaml:644]: I don't insist on you adding it in this PR, but the dispatch project is missing a `.dockerignore` file. This means that people who have a `build` directory in their checkout and want to build the image will suffer the time penalty while docker copies the entire directory (with all logs from the test runs, ...) into the docker build context.

Good .dockerignore for dispatch can look something like this, https://github.com/skupperproject/skupper-router/blob/9017643e419af523236f643d54b0aca42f109e57/.dockerignore

Comment (Jmennius) [reply .github/workflows/build.yaml:644]: Added, hope it's fine.

**Comment by jiridanek** [comment]
@ganeshmurthy @kgiusti I'm going to merge this if there aren't any more comments. Thanks.

**Review by kgiusti — APPROVED** [review]
@ganeshmurthy @jiridanek looks good to me.  Feel free to commit it if you're both happy.

**Review by ganeshmurthy — APPROVED** [review]

**Pull request merged**
